### PR TITLE
[SPARK-21388][ML][PySpark] GBTs inherit from HasStepSize & LInearSVC from HasThreshold

### PR DESCRIPTION
**Pull request by zhengruifeng**
## What changes were proposed in this pull request?
GBTs inherit from HasStepSize & LInearSVC/Binarizer from HasThreshold

## How was this patch tested?
existing tests

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/param/shared/sharedParams.scala [hunk 1]
```diff
@@ -162,7 +162,7 @@ private[ml] trait HasThreshold extends Params {
    * Param for threshold in binary classification prediction, in range [0, 1].
    * @group param
    */
-  final val threshold: DoubleParam = new DoubleParam(this, "threshold", "threshold in binary classification prediction, in range [0, 1]", ParamValidators.inRange(0, 1))
+  val threshold: DoubleParam = new DoubleParam(this, "threshold", "threshold in binary classification prediction, in range [0, 1]", ParamValidators.inRange(0, 1))
 
   setDefault(threshold, 0.5)
```
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/param/shared/sharedParams.scala]: Maybe the error can be eliminated after removing this line, and actually we should set default value in the concrete class if it's inherited.

**Comment by holdenk** [comment]
So improving the doc certainly makes sense on the Python side (thank you for catching that). But as @MLnick asked on the JIRA - I'm not super sure I see the point of the refactor on the Scala side. If we have a parameter with the same name but a different meaning why mix in a base trait only to override its only meaningful provided value?

What do you think @jkbradley?

**Comment by zhengruifeng** [comment]
@holdenk I think the meaning of `StepSize` in GBT and `Threshould` in LInearSVC/Binarizer is almost same as that in other algs, so it maybe better to make them inherit from same trait?

**Comment by MLnick** [comment]
It makes sense for `LinearSVC` since the threshold has the same meaning for other binary classifiers (e.g. LogisticRegression). The meaning of threshold for `Binarizer` is very different as it operates on feature values in a feature vector, not raw predicted values from a model. I would say back out the Binarizer changes here.

As for GBT, well I am sort of indifferent as the step size is roughly analogous between the two definitions (in both cases it is a learning rate).

File: python/pyspark/ml/feature.py [hunk 2]
```diff
@@ -61,7 +61,8 @@
 
 
 @inherit_doc
-class Binarizer(JavaTransformer, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):
+class Binarizer(JavaTransformer, HasInputCol, HasOutputCol, HasThreshold,
```
Comment (MLnick) [inline python/pyspark/ml/feature.py]: This also needs to be reverted.

File: python/pyspark/ml/feature.py [hunk 3]
```diff
@@ -108,20 +109,6 @@ def setParams(self, threshold=0.0, inputCol=None, outputCol=None):
         kwargs = self._input_kwargs
         return self._set(**kwargs)
 
-    @since("1.4.0")
```
Comment (MLnick) [inline python/pyspark/ml/feature.py]: and similarly here, added back

**Review by yanboliang — APPROVED** [review]
LGTM

**Comment by yanboliang** [comment]
Merged into master, thanks for all.

**Pull request closed without merging**